### ISSUE TITLE
"plugin"/hooking system

### ISSUE THREAD

**Issue by atticus-sullivan** [atticus-sullivan/cluttealtex#25]
Left over stuff from https://github.com/atticus-sullivan/cluttealtex/issues/17

Also related (but not part of this issue #15

----

In order to keep `typeset` a bit smaller, the plan is to implement a hooking mechanism. Currently I think of two places that can be hooked:

1. for running en external command after everything tex related (including recover) is done (think of `sagetex`, `memoize`, `biber` , `makeindex` etc)
2. for adding code to `tex_inject` (think of `memoize`, `quiet`, `epstopdf`, `minted`)

Idea: Make a space in the `options` to register these hooks in a `priority -> function` table (would be cool to ensure no hook is being overwritten). When the user sets an option, the corresponding `handle_{cli,cfg}` function (defined in `option_spec`) can then add a reference to the corresponding function (in `typeset_hooks`) to the corresponding hooking table in `options`.

Still open: How handle hooks like `epstopdf` or `minted` which seem to be used automatically (at least the `\PassOptionsToPackage` in the inject hook)? (there is no `handle_{cli,cfg}` that adds these hooks. Should they be registered by default or only when detected?

---

Also regarding refactoring: for `minted` and `epstopdf` the support is enabled via `--package-support=minted`/`--package-support=epstopdf` this seems to be inconsistent with `biber`/`sagetex`/`memoize`.
While it might not be possible to enable the support for other packages via `--package-support` as some of these arguments take a parameter, maybe we should add a `--minted`/`--epstopdf` option to be consistent.

---

Suggestion:

Should we in a plugin manner try to move everything related to advanced options like `glossaries`/`memoize` (mainly stuff that needs more complex hooking functions) to dedicated files that return a table for use in the `option_spec`? This would group the complete implementation of the option in one file instead of spreading it over multiple files (`option_spec` due to the `handle_` function(s) and `typeset_hooks` due to the hooking).

 => I like this idea.

Also an idea would be to try to merge the help output with the option_spec (further reduce the places for modifications needed in when adding a new option). The concrete help output would then be generated dynamically by iterating over the `option_spec` (we'd need an ordering key in order to specify a deterministic ordering of the help output).

---

The introduction of the `handle_{cfg,cli}` function makes the code more concise and groups the code related to a particular option in the `option_spec`. On the other hand, this may make the codebase harder to understand for new developers. Writing some developer documentation for this aspect (as well as the hooking mechanism that might be introduced in the future) probably will prove most valuable.

---

- [x] Verbose should show what hooks are executed

---

Keep minted etc inconsistency (note this stems from the old days and it's not directly planned to adjust, maybe open an issue with won't fix where users can suggest / offer help)
basically the issue and everything is the same with memoize, but I decided to run the shell command myself (no shell enable necessary and more control) instead like with minted let tex run the command.

---

supported packages:
- make accumulating
- table input
- only activate when flag is passed? (emit warning if not and use is detected)
- add pred_suggestion to hooks?

**Comment by atticus-sullivan** [comment]
1. add hooking fields to option type
2. execute + log the hooks
3. write example function for one hook
4. install hook when option is passed
5. write other hooks

---

- for later: move option type (spec + options) to option_spec?
